### PR TITLE
Feature: honor requirements versions

### PR DESCRIPTION
**Pull request by sobregosodd**
closes issue https://github.com/DataDog/guarddog/issues/362
closes issue https://github.com/DataDog/guarddog/issues/262
closes issue https://github.com/DataDog/guarddog/issues/86

**Context**
Guarddog verify command operated against all possible matches of packages. 
So for instance a requirement with `guarddog>=1.0.0`
would result in guarddog verifying all possible matches:
```- 1.0.0
- 1.0.1
- 1.0.2
- ... 
```
This behaviour could lead to a very lengthy scan process and most-likely not the most useful given that only the most recent match is installed by package managers
 
**Changes**
This PR modifies that behaviour to select by default the best "match" (like a real package manager would do)
So, having the following requirement `guarddog>=1.0.0`
Would result for selecting the latest matched version, e.g: `1.9.0`

IMPORTANT: This default behaviour can be overiden setting the environment flag `GUARDDOG_VERIFY_EXHAUSTIVE_DEPENDENCIES` to "True"

### CODE REVIEW

**Review by nnicogomez — APPROVED** [review]

File: guarddog/scanners/pypi_project_scanner.py [hunk 1]
```diff
@@ -97,49 +102,26 @@ def safe_parse_requirements(req):
...
+                    )
+                    project_exists_on_pypi = False
+                    continue
+
+                r = Requirement(str(requirement))
+
+                matched_versions = [m for m in r.specifier.filter(available_versions)]
+
+                if not VERIFY_ALL_DEPENDENCIES and matched_versions:
+                    matched_versions = [sorted(matched_versions).pop()]
```
Comment (christophetd) [inline guarddog/scanners/pypi_project_scanner.py]: aren't you missing the last version due to the pop()? I think (not sure) that this should be:

```
matched_versions.extend(sorted(matched_versions))
```

Comment (sobregosodd) [reply guarddog/scanners/pypi_project_scanner.py]: No, is the other way around, we're only keeping the last one.
i.e:
```
>>> x = {2,3,1}
>>> sorted(x)
[1, 2, 3]
>>> sorted(x).pop()
3
```

This is executed when `VERIFY_ALL_DEPENDENCIES` is **false**

Comment (christophetd) [reply guarddog/scanners/pypi_project_scanner.py]: are we sure that `sorted` is the appropriate way to sort version strings?

Comment (sobregosodd) [reply guarddog/scanners/pypi_project_scanner.py]: No:
```

>>> from semantic_version import Version # npm
>>> vmax = "1.10.0"
>>> vmin = "1.2.0"
>>> vmax > vmin
False
>>> Version(vmax) > Version(vmin)
True
...

>>> from packaging.version import Version # pypi
>>> Version(vmax) > Version(vmin)
True
```

I can address that by sorting the keys with the semver specs

Comment (sobregosodd) [reply guarddog/scanners/pypi_project_scanner.py]: fixed in a99ed28495f1bcf111324fcb52e4cd0e074bb8c4

**Review by christophetd — APPROVED** [review]
As a last comment, it would be good to add debug output to show which package version precisely we're scanning

**Comment by sobregosodd** [comment]
> As a last comment, it would be good to add debug output to show which package version precisely we're scanning

Not following, the package version is expressively being printed out to console (also included in the json format), for example:
```
Scanning using at most 10 parallel worker threads
Found 0 potentially malicious indicators scanning Django version 3.2.15
```

**Pull request merged**